state-chain/pallets/cf-ingress-egress/src/lib.rs

```diff
@@ -1965,6 +1969,284 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
+		deposit_address: TargetChainAccount<T, I>,
+		source_address: Option<ForeignChainAddress>,
+		action: ChannelAction<T::AccountId>,
+		broker: &T::AccountId,
+		boost_fee: u16,
+		boost_status: BoostStatus<TargetChainAmount<T, I>>,
+		channel_id: u64,
+		block_height: TargetChainBlockNumber<T, I>,
+		origin: DepositOrigin,
+	) -> Option<BoostStatus<TargetChainAmount<T, I>>> {
+		if amount < MinimumDeposit::<T, I>::get(asset) {
+			// We do not process/record pre-witnessed deposits for amounts smaller
+			// than MinimumDeposit to match how this is done on finalisation
+			return None;
+		}
+
+		if let Some(tx_id) = deposit_details.deposit_id() {
+			if TaintedTransactions::<T, I>::mutate(broker, &tx_id, |opt| {
+				match opt.as_mut() {
+					// Transaction has been reported, mark it as pre-witnessed.
+					Some(status @ TaintedTransactionStatus::Unseen) => {
+						*status = TaintedTransactionStatus::Prewitnessed;
+						true
+					},
+					// Pre-witnessing twice is unlikely but possible. Either way we don't want
+					// to change the status and we don't want to allow boosting.
+					Some(TaintedTransactionStatus::Prewitnessed) => true,
+					// Transaction has not been reported, mark it as boosted to prevent further
+					// reports.
+					None => false,
+				}
+			}) {
+				return None;
+			}
+		}
+
+		let prewitnessed_deposit_id = PrewitnessedDepositIdCounter::<T, I>::mutate(|id| -> u64 {
+			*id = id.saturating_add(1);
+			*id
+		});
+
+		// Only boost on non-zero fee and if the channel isn't already boosted:
+		if T::SafeMode::get().boost_deposits_enabled &&
+			boost_fee > 0 &&
+			!matches!(boost_status, BoostStatus::Boosted { .. })
+		{
+			match Self::try_boosting(asset, amount, boost_fee, prewitnessed_deposit_id) {
+				Ok(BoostOutput { used_pools, total_fee: boost_fee_amount }) => {
+					let amount_after_boost_fee = amount.saturating_sub(boost_fee_amount);
+
+					// Note that ingress fee is deducted at the time of boosting rather than the
+					// time the deposit is finalised (which allows us to perform the channel
+					// action immediately):
+					let AmountAndFeesWithheld { amount_after_fees, fees_withheld: ingress_fee } =
+						Self::conditionally_withhold_ingress_fee(
+							asset,
+							amount_after_boost_fee,
+							&origin,
+						);
+
+					let used_pool_tiers = used_pools.keys().cloned().collect();
+
+					if let Ok(action) = Self::perform_channel_action(
+						action,
+						asset,
+						source_address,
+						amount_after_fees,
+						origin.clone(),
+					) {
+						Self::deposit_event(Event::DepositBoosted {
+							deposit_address,
+							asset,
+							amounts: used_pools,
+							block_height,
+							prewitnessed_deposit_id,
+							channel_id,
+							deposit_details: deposit_details.clone(),
+							ingress_fee,
+							boost_fee: boost_fee_amount,
+							action,
+							origin_type: origin.into(),
+						});
+					} else {
+						// TODO: emit error?
```

Comment:
I looked into how this can fail, and it looks like it can only happen when `try_credit_account` fails, which itself only fails when balance overflows... Maybe we could just make `try_credit_account` infallible instead? We don't really have any place where this is handled by the caller in a meaningful way: we either unwrap, print error or propogate it to the point where we no longer know what to do with it and simply ignore (as has been the case here even prior to this PR). Maybe we could just make it infallible and either saturate and/or use `log_or_panic` internally? Would simplify a lot of code. @dandanlen

Reply:
Yeah I agree - I think assuming we can't overflow is almost always a fair assumption. I think it was left as fallible because of uncertainty in the AMM implementation (eg. at the extreme ends of possible pools state / swap rates). So as a compromise I would suggest creating a separate infallible (saturating) method, and then use the fallible one when crediting AMM funds, and the saturating one when crediting funds from deposits, for example .

Reply:
OK, though I still don't quite understand the scenario where it would overflow. For AMM to return an amount that might overflow, it has to exist in the pool in the first place... What am I missing?